File: MATTHEW/leetcode/AIcoding/1.Gemini-2.5/erlang_solutions/173-binary-search-tree-iterator.erl
-module(bst_iterator).

-export([new/1, next/1, has_next/1]).

new(Root) ->
    push_left_spine(Root, []).

next([#{val := Val, right := RightChild} | RestStack]) ->
    NewStack = push_left_spine(RightChild, RestStack),
    {Val, NewStack}.

has_next([]) ->
    false;
has_next(_) ->
    true.

push_left_spine(nil, Stack) ->
    Stack;
push_left_spine(#{left := LeftChild} = Node, Stack) ->
    push_left_spine(LeftChild, [Node | Stack]).

File: leetcode/AIcoding/3.DeepSeek-v3/c_solutions/173-binary-search-tree-iterator.c
#include <stdlib.h>

typedef struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
} TreeNode;

typedef struct {
    TreeNode** stack;
    int top;
} BSTIterator;

BSTIterator* bSTIteratorCreate(struct TreeNode* root) {
    BSTIterator* iterator = (BSTIterator*)malloc(sizeof(BSTIterator));
    iterator->stack = (TreeNode**)malloc(100000 * sizeof(TreeNode*));
    iterator->top = -1;

    TreeNode* current = root;
    while (current != NULL) {
        iterator->stack[++(iterator->top)] = current;
        current = current->left;
    }

    return iterator;
}

int bSTIteratorNext(BSTIterator* obj) {
    TreeNode* node = obj->stack[(obj->top)--];
    int result = node->val;

    node = node->right;
    while (node != NULL) {
        obj->stack[++(obj->top)] = node;
        node = node->left;
    }

    return result;
}

bool bSTIteratorHasNext(BSTIterator* obj) {
    return obj->top >= 0;
}

void bSTIteratorFree(BSTIterator* obj) {
    free(obj->stack);
    free(obj);
}